### ISSUE TITLE
Cannot reference outer query when using lateral join and nested correlated subqueries

### ISSUE THREAD

**Issue by xiasongh** [trinodb/trino#23822]
I found that when using `LATERAL`, I cannot reference the outer query from a nested correlated subquery. DuckDB is able to do so, is this currently not supported or do I need to write the query in some specific way for this to work?

I've provided an example query
```
select * from (
values
    timestamp '2024-01-01',
    timestamp '2024-02-01',
    timestamp '2024-03-01'
) as timestamps(ts)
left join lateral (
    with past_table as (
        select * from changelog
        join (
            select id, max(ts) as ts
            from changelog
            where ts <= timestamps.ts
            group by pk
        ) using (pk, ts)
    )

    select count(*) as my_count from past_table
) on true;
```

To provide a bit of context on the query itself (though I think it's not necessarily relevant): I have a changelog that is collected from Postgres' WAL and I want to perform a time travel query for multiple points in time. In the outer query, I have the timestamps of interest, and for each timestamp, I want to reconstruct the table's past state, then run some query but with the past table.

I found some issues that don't use `LATERAL` but may be the same root cause
https://github.com/trinodb/trino/issues/14199
https://github.com/trinodb/trino/issues/18727

**Comment by hashhar** [comment]
cc: @kasiafi ?

**Comment by kasiafi** [comment]
@xiasongh you're right that your issue and the other issues you mentioned have the same root cause. It is Trino's limited capability to de-correlate subqueries. In your case, it cannot de-correlate the nested join. Additionally, it cannot deal with a combination of the aggregation (`max(ts) group by pk`) with a non-equality predicate `ts <= timestamps.ts`. 


I tried to rewrite your query so that it is not correlated (excuse my pseudocode)
```
LEFT JOIN USING (changelog.pk, ts)
    - SELECT timestamps.ts, changelog.pk, max(changelog.ts) as ts 
      FROM 
        LEFT JOIN ON (timestamps.ts >= changelog.ts)
            - timestamps(ts)
            - changelog(ts, pk)
        GROUP BY (timestamps.ts, changelog.pk)
    - changelog(ts, pk)              
```

If you're interested in de-correlation limitations, see this comment: https://github.com/trinodb/trino/issues/14199#issuecomment-2483490115
